File: iosApp/iosApp/Modules/ImprovementProposals/ImprovementProposalsWireframeFactory.swift
// Created by web3d3v on 30/07/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3walletcore

protocol ImprovementProposalsWireframeFactory {
    func make(_ parent: UIViewController?) -> ImprovementProposalsWireframe
}

final class DefaultImprovementProposalsWireframeFactory {
    private let improvementProposalWireframeFactory: ImprovementProposalWireframeFactory
    private let improvementProposalsService: ImprovementProposalsService

    init(
        improvementProposalWireframeFactory: ImprovementProposalWireframeFactory,
        improvementProposalsService: ImprovementProposalsService
    ) {
        self.improvementProposalWireframeFactory = improvementProposalWireframeFactory
        self.improvementProposalsService = improvementProposalsService
    }
}

extension DefaultImprovementProposalsWireframeFactory: ImprovementProposalsWireframeFactory {

    func make(_ parent: UIViewController?) -> ImprovementProposalsWireframe {
        DefaultImprovementProposalsWireframe(
            parent,
            improvementProposalWireframeFactory: improvementProposalWireframeFactory,
            improvementProposalsService: improvementProposalsService
        )
    }
}

final class ProposalsWireframeFactoryAssembler: AssemblerComponent {

    func register(to registry: AssemblerRegistry) {
        registry.register(scope: .instance) { resolver -> ImprovementProposalsWireframeFactory in
            DefaultImprovementProposalsWireframeFactory(
                improvementProposalWireframeFactory: resolver.resolve(),
                improvementProposalsService: resolver.resolve()
            )
        }
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/improvementproposals/ImprovementProposalsWireframeFactory.kt
package com.sonsofcrypto.web3wallet.android.modules.improvementproposals

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.AssemblerComponent
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistry
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistryScope
import com.sonsofcrypto.web3wallet.android.modules.compose.improvementproposal.ImprovementProposalWireframeFactory
import com.sonsofcrypto.web3walletcore.modules.improvementProposals.ImprovementProposalsWireframe
import com.sonsofcrypto.web3walletcore.services.improvementProposals.ImprovementProposalsService
import smartadapter.internal.extension.name

interface ImprovementProposalsWireframeFactory {
    fun make(parent: Fragment?): ImprovementProposalsWireframe
}

class DefaultImprovementProposalsWireframeFactory(
    private val improvementProposalsService: ImprovementProposalsService,
    private val improvementProposalWireframeFactory: ImprovementProposalWireframeFactory,
): ImprovementProposalsWireframeFactory {

    override fun make(parent: Fragment?): ImprovementProposalsWireframe {

        return DefaultImprovementProposalsWireframe(
            parent?.let { WeakRef(it) },
            improvementProposalsService,
            improvementProposalWireframeFactory,
        )
    }
}

class ImprovementProposalsWireframeFactoryAssembler: AssemblerComponent {

    override fun register(to: AssemblerRegistry) {

        to.register(ImprovementProposalsWireframeFactory::class.name, AssemblerRegistryScope.INSTANCE) {
            DefaultImprovementProposalsWireframeFactory(
                it.resolve(ImprovementProposalsService::class.name),
                it.resolve(ImprovementProposalWireframeFactory::class.name),
            )
        }
    }
}